library/src/meta/figure.rs

```diff
@@ -32,41 +104,230 @@ pub struct FigureElem {
+            }
+
+            let number = element
+                .counter
+                .at(vt, self.0.location().expect("missing location"))?
+                .display(vt, &element.numbering)?
+                .spanned(self.span());
+
+            Ok(Some(name + number))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Builds the caption for the figure.
+    /// If there is a numbering, will also try to show the supplement and the numbering.
+    ///
+    /// # Errors
+    /// If a numbering is specified but the [`Self::element`] is `None`.
+    pub fn show_caption(&self, vt: &mut Vt, styles: StyleChain) -> SourceResult<Content> {
+        let Some(mut caption) = self.caption(styles) else {
+            return Ok(Content::empty());
+        };
+
+        if let Some(sup_and_num) = self.show_supplement_and_numbering(vt, None)? {
+            caption = sup_and_num + TextElem::packed(": ") + caption;
+        }
+
+        Ok(caption)
+    }
 }
 
 impl Synthesize for FigureElem {
-    fn synthesize(&mut self, styles: StyleChain) {
+    fn synthesize(&mut self, vt: &mut Vt, styles: StyleChain) -> SourceResult<()> {
         self.push_numbering(self.numbering(styles));
+
+        // We get the numbering or `None`.
+        let numbering = self.numbering(styles);
+        let supplement = self.supplement(styles);
+
+        // We get the content or `None`.
+        let content = match self.kind(styles) {
+            Smart::Auto => match self.determine_type(styles, supplement.is_auto()) {
+                Some(ty) => Some(ty),
+                None => bail!(
+                    self.span(),
+                    "unable to determine figure type, use `kind` to manually specify it"
+                ),
+            },
+            Smart::Custom(ContentParam::Elem(ty)) => self.find_elem(ty),
+            Smart::Custom(ContentParam::Name(_)) => None,
+        };
+
+        if self.kind(styles).is_auto() {
+            if let Some(content) = &content {
+                self.push_kind(Smart::Custom(ContentParam::Elem(content.func())));
+            }
+        }
+
+        // The list of choices is the following:
+        // 1. If there is a detected content, we use the counter `counter(figure.where(kind: detected_content))`
+        // 2. If there is a name/elem, we use the counter `counter(figure.where(kind: name/elem))`
+        // 4. We return None.
+        let counter = if let Some(content) = &content {
+            Some(Counter::new(CounterKey::Selector(Selector::Elem(
+                Self::func(),
+                Some(dict! {
+                    "kind" => Value::from(content.func()),
+                }),
+            ))))
+        } else if let Smart::Custom(content) = self.kind(styles) {
+            Some(Counter::new(CounterKey::Selector(Selector::Elem(
+                Self::func(),
+                Some(dict! {
+                    "kind" => Value::from(content),
+                }),
+            ))))
```

Comment:
this is quite duplicated

Reply:
Cleaned it up nicely by adding a `as_custom` to `Smart` making it return an `Option` if `self` is `Smart::Custom(c)`